include/swift/Basic/Located.h

```diff
@@ -0,0 +1,50 @@
+//===--- Located.h - Source Location and Associated Value ----------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Provides a currency data type Located<T> that should be used instead of std::pair<T, SourceLoc>.
+//
+//===----------------------------------------------------------------------===//
+
+
+#ifndef SWIFT_BASIC_LOCATED_H
+#define SWIFT_BASIC_LOCATED_H
+#include "swift/Basic/SourceLoc.h"
+
+namespace swift {
+
+/// A currency type for keeping track of items which were found in the source code.
+///
+/// Several parts of the compiler need to keep track of a `SourceLoc` corresponding to an item, in case
+/// they need to report some diagnostics later. For example, the ClangImporter needs to keep track
+/// of where imports were originally written. Located makes it easy to do so while making the code
+/// more readable, compared to using `std::pair`.
+template<typename T>
+struct Located {
+
+  /// The main item whose source location is being tracked.
+  T item;
+
+  /// The original source location from which the item was parsed.
+  SourceLoc loc;
+
+  Located() {}
```

Comment:
Here, I had to create default constructor because currently we are using these Located in DenseMap and there are code like
```c++

  value_type& FindAndConstruct(const KeyT &Key) {
    BucketT *TheBucket;
    if (LookupBucketFor(Key, TheBucket))
      return *TheBucket;

    return *InsertIntoBucket(TheBucket, Key);
  }

  ValueT &operator[](const KeyT &Key) {
    return FindAndConstruct(Key).second; /**Here**/
  }
```
/**Here**/ There is a call to default constructor. Should we make any workarounds around that @varungandhi-apple ?

Reply:
Thanks for drawing attention to it, I missed that. Shouldn't the default constructor call the default constructors of the fields? My C++ initialization-fu isn't that great but I think it should instead be something like `Located() : item(), loc() {}` -- if we don't do that, I'm guessing we risk having the fields be uninitialized?

Reply:
Yes, I agree this should be called otherwise it will be undefined behaviour. Thanks.